pyproject.toml

```diff
@@ -183,6 +183,35 @@ features = ['docs']
 build = "cd docs && make html"
 serve = "sphinx-autobuild docs docs/_build --host 0.0.0.0"
 
+[tool.hatch.envs.upstream]
+dependencies = [
+    'numpy>=2.2.0.dev0',
+    'numcodecs @ git+https://github.com/zarr-developers/numcodecs',
+    'fsspec @ git+https://github.com/fsspec/filesystem_spec',
+    's3fs @ git+https://github.com/fsspec/s3fs',
+    'universal_pathlib @ git+https://github.com/fsspec/universal_pathlib',
+    'crc32c @ git+https://github.com/ICRAR/crc32c',
+    'typing_extensions @ git+https://github.com/python/typing_extensions',
+    'donfig @ git+https://github.com/pytroll/donfig',
+    # test deps
+    'hypothesis',
+    'pytest',
+    'pytest-cov',
+    'pytest-asyncio',
+    'moto[s3]',
+]
+
+# help! this is not working for numpy :(
+[tool.hatch.envs.upstream.env-vars]
+PIP_EXTRA_INDEX_URL = "https://pypi.anaconda.org/scientific-python-nightly-wheels/simple"
+PIP_PRE = "1"
```

Comment:
I could use some help here. Maybe @d-v-b, @dcherian, or @jakirkham have ideas.

This index is what we want to be using for NumPy (and eventually numcodecs) but something is not right. I can only get this to work if I also a lower bound (`>=2.2.0.dev0`) to the numpy dependency.

Reply:
https://scientific-python.org/specs/spec-0004/#test-with-nightly-wheels suggests explicitly installing the development dependencies like

```
python -m pip install --pre --upgrade --extra-index-url https://pypi.anaconda.org/scientific-python-nightly-wheels/simple numpy scipy
```

Reply:
Yep. I read that. My goal here is to put configure everything in our hatch env configuration but if we can't get that to work, we may have to do some extra installs in the CI action itself.

Reply:
Actually, [4c9619b](https://github.com/zarr-developers/zarr-python/pull/2418/commits/4c9619b29a95adf5fafcfefcaa8d48cb11ddf267) seems to have worked!